Abstraction for morphology collection.

1uc opened the pull request:
Currently, MorphIO requires the file, filename or HDF5 group which contains the morphology. This PR adds an interface for loading morphologies by name from a `Collection` object.

The first commits improve style, turn off problematic flags, add missing ctors and fix a bug in the HDF5 reader. Furthermore, they add gulrak's implementation of `std::filestystem`. The last commit adds the proposed feature and its tests.

1uc commented:
Please let me know if you want to split certain commits into their own PRs.

1uc commented:
This would implement part of #443.

mgeplf commented on include/morphio/collection.h:
```diff
@@ -0,0 +1,82 @@
+namespace morphio {
+
+class CollectionImpl
+{
```
can we just forward decl this, and move the `load`s to `collection.cpp`?  I *try* (often unsuccessfully) to keep the headers as simple as possible.

1uc replied:
Good point. Solved via explicit instantiation.

mgeplf commented on include/morphio/collection.h:
```diff
@@ -0,0 +1,82 @@
+     * point to the directory containing the morphology files.
+     *
+     * If the collection path is a directory, the extension of the morphology
+     * file must be guessed. The optional argument `extensions` specifies which
```
nice api; good idea

mgeplf commented on include/morphio/collection.h:
```diff
@@ -0,0 +1,82 @@
+     * and in which order the morphologies are searched.
+     */
+    Collection(std::string collection_path,
+               std::vector<std::string> extensions = std::vector<std::string>{"h5", "swc", "asc"});
```
at bbp, I think we should put `swc` last; since it's generally not used.  Annoyingly, I'm not sure we're 100% standardized on lowercase `asc`, so adding `ASC` might be required :(  I will look around.

mgeplf commented on include/morphio/mut/morphology.h:
```diff
@@ -41,7 +41,10 @@ class Morphology
     **/
     explicit Morphology(const std::string& uri, unsigned int options = NO_MODIFIER);
 
-    /// Build a mutable Morphology from a mutable morphology aaaaaa
```
ah yes, the classic `aaaaaa` :facepalm:

mgeplf commented on src/collection.cpp:
```diff
@@ -0,0 +1,126 @@
+    std::vector<std::string> _extensions;
+};
+
+class ContainerCollection: public morphio::detail::CollectionImpl<ContainerCollection>
```
could we add HDF5 to the name, just so it's clear what the backing store is?

1uc replied:
done.

mgeplf commented on src/readers/morphologyHDF5.cpp:
```diff
@@ -318,13 +318,13 @@ void MorphologyHDF5::_read(const std::string& groupName,
+    const auto group = groupName == "" ? _group : _group.getGroup(groupName);
 
-    if (!_group.exist(groupName)) {
+    if (!group.exist(datasetName)) {
```
is this the bug you reference in the PR description?

1uc replied:
Yes, when adding a test with a container that contained morphologies with perimeters, it would complain that there was to dataset called `"perimeters"`.

What makes me slightly wary is that we've used containers in TouchDetector. However, it's not impossible, that we've simply never used more complex morphologies with a perimeter.

The proposed change I believe is correct, whether the previous was a bug I'm not 100% sure. I strongly suspect it was, but as mentioned it would have been quite good at hiding.

1uc replied:
Sorry! I've not reread my description. Yes, the `datasetName` is relative to `groupName`. For fancy morphologies the `groupName` might be something like `"organelles/mitochondria"` and the `datasetName` would be `"points"`.

Hence, if we want to check that the dataset exists, we need to check the `group` (not `_group`) and check `datasetName`. It's highly likely it's a copy-pasting bug from the check on line 321.

mgeplf reviewed (commented):
Looks good to me, I need to find some time to try it, but I think w/ a couple small changes, we'll be ready to merge.

mgeplf commented on include/morphio/collection.h:
```diff
@@ -0,0 +1,82 @@
+     * The object should not be used after calling `close`.
+     */
+    void close();
+
```
I wonder if a useful addition would be to have `list` or `contents` to get the available morphologies within the container?

1uc replied:
Good idea! I was thinking we'd get the morphology names from libsonata via a nodesfile; but that's not always the case. For directories it might be difficult to distinguish between files that contain morphologies and files that happen to have an extension that's used for morphologies. This gets slightly murkier since we often(?) nest morphologies in subdirectories of the morphology directory. The subdirectory then becomes part of the morphology name.

1uc replied:
@mgeplf Can this please be a separate task? I'd like to prepare the migration for two of our internal applications; but their CI will only run after this has been merged. Therefore, it would be helpful to limit the functionality of this PR.

So far I've not needed know what's inside a collection, I've always had a nodesfile tell me what I need to load.

mgeplf replied:
yeah, that's no problem

We'll need a way to list what's there, at least of h5 files, so that we can quickly reconcile that with nodes files, if things aren't working - but we can do that in the next step.

1uc commented on CMakeLists.txt:
```diff
@@ -7,6 +7,7 @@ set(CMAKE_VERBOSE_MAKEFILE ON)
 
 option(BUILD_BINDINGS "Build the python bindings" ON)
 option(MorphIO_CXX_WARNINGS "Compile C++ with warnings" ON)
+option(MorphIO_WERROR "If compiled with warnings, additionally activate '-Werror'" OFF)
```
That would lead to some double negation logic. We could just switch the default to `On`. Either is fine, as long as I can compile with warnings without them preventing progress by turning into errors.

1uc replied:
If `*_NO_WERROR` is your preferred name, I'm happy to change it. For now I only updated the default.

1uc commented:
Thank you for the review. I'll change the unaddressed comments as suggested.

Currently, I'm working on integrating this into TouchDetector. We can try this "new" API (it's mostly copied from an internal library) there before merging it. This would check the C++ API and currently missing thread-safety (WIP).

After that we'll also need to prepare SpatialIndex. Which would check that the Python API is fine.

mgeplf commented:
I'm basically happy with this.  Is there anyone else you'd like to review it?

1uc commented:
I've run it through TD and that works. I'm checking SI.

1uc commented on CMake/CompilerFlags.cmake:
```diff
@@ -12,7 +12,10 @@ if (WIN32)
   # Taken from https://github.com/BlueBrain/hpc-coding-conventions/blob/master/cpp/cmake/bob.cmake#L192-L255
   if(${PROJECT_NAME}_CXX_WARNINGS)
-    set(FLAGS "${FLAGS} -Werror -Weverything")
+    if(${PROJECT_NAME}_WERROR)
```
I think it's good to be able to see warnings while developing. Often they're just that warnings, some are true some are false. Yes, one should work towards eliminating them all; otherwise one misses the important one.

My issue with `-Werror` is that it makes appeasing the compiler priority number one. Nothing else is possible until there are no more warnings.

Concretely, GCC 12.2.1 will emit many warnings from within third-party libraries.

mgeplf replied:
Ok, you've convinced me; we'll keep it.  Thanks.

mgeplf commented:
I've tried merging this several times, but w/ the github outage, I think it's best I wait until tomorrow.
